promsonnet/prom.libsonnet

```diff
@@ -0,0 +1,39 @@
+{
+  v1: {
+    ruleGroupSet: {
+      new():: {
+        groups_map:: {},
+        groups_order:: [],
+        local groups_map = self.groups_map,
+        local groups_order = self.groups_order,
+        groups: [groups_map[group] for group in groups_order],
+      },
+      addGroup(group):: {
+        groups_map+:: {
+          [group.name]: group,
+        },
+        groups_order+:: [group.name],
+      },
+    },
+
+    ruleGroup: {
+      new(name):: {
+        name: name,
+        rules_map:: {},
+        rules_order:: [],
+        local rules_map = self.rules_map,
+        local rules_order = self.rules_order,
+        rules: [rules_map[rule] for rule in rules_order],
+      },
+
+      rule: {
+        new(name, rule):: {
```

Comment:
How about having a `newAlert` and a `newRecording` constructor instead of this `new`? They could then verify that only the allowed fields are set, and they could provide a default naming from the `name` key to not accidentally have them inconsistent.

`newAlert` would set the `alert` field from the `name` key (but give an explicitly specified field precedence)p

`newRecording` would set the `record` field from the `name` key (but give an explicitly specified field precedence).

Reply:
I'm not sure how we'd do validation - that's not something Jsonnet naturally lends itself to. However, a `newAlert` and `newRecording` would allow for setting the correct `name` fields as you suggest.

Reply:
done as suggested.